### ISSUE TITLE
Nightscout Support

### ISSUE THREAD

**Issue by Klaus3d3** [AmazMod/AmazMod#7]
Hi all,
I've already talked to Pavel from Huami Intl.
I'm a diabetic and using the app xdrip+ to monitor my blood glucose level. 
Since now i've used an android wear watch or a pebble to show my level on my watch. I'd like to have this on my stratos, too. I know it's very difficult or nearly impossible to have this in a watchface, but having a springboard page/widget should be possible.
The main problem i have, is getting the data to the watch. Pavel told me, your could eventually help.
Xdrip provides the data via a simple local web service on the phone under: http://127.0.0.1:17580/sgv.json woth data like this:
```json
[  
   {  
      "_id":"6d2cd8cb-efdb-461a-9a42-9355aef68f4e",
      "device":"LimiTTer",
      "dateString":"2018-04-04T21:23:16.436+0200",
      "sysTime":"2018-04-04T21:23:16.436+0200",
      "date":1522869796436,
      "sgv":78,
      "delta":-1.102,
      "direction":"Flat",
      "noise":1,
      "filtered":103764,
      "unfiltered":103764,
      "rssi":100,
      "type":"sgv",
      "units_hint":"mgdl"
   },
   {  
      "_id":"dd12ddda-ae6f-43c7-bede-3243d204794b",
      "device":"LimiTTer",
      "dateString":"2018-04-04T21:17:16.436+0200",
      "sysTime":"2018-04-04T21:17:16.436+0200",
      "date":1522869436436,
      "sgv":79,
      "delta":24.693,
      "direction":"DoubleUp",
      "noise":1,
      "filtered":105294,
      "unfiltered":105294,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"29fb85b4-d1d6-44dc-95ec-842f90e31fca",
      "device":"LimiTTer",
      "dateString":"2018-04-04T21:12:38.475+0200",
      "sysTime":"2018-04-04T21:12:38.475+0200",
      "date":1522869158475,
      "sgv":56,
      "delta":10.575,
      "direction":"SingleUp",
      "noise":1,
      "filtered":78823,
      "unfiltered":78823,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"22d409d3-4c87-4c15-a1f7-4c9436114980",
      "device":"LimiTTer",
      "dateString":"2018-04-04T21:07:38.475+0200",
      "sysTime":"2018-04-04T21:07:38.475+0200",
      "date":1522868858475,
      "sgv":46,
      "delta":-15.761,
      "direction":"SingleDown",
      "noise":1,
      "filtered":66588,
      "unfiltered":66588,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"3ad034a0-9da3-4eae-9dd0-b6a1f3695592",
      "device":"LimiTTer",
      "dateString":"2018-04-04T21:02:38.475+0200",
      "sysTime":"2018-04-04T21:02:38.475+0200",
      "date":1522868558475,
      "sgv":61,
      "delta":5.084,
      "direction":"FortyFiveUp",
      "noise":1,
      "filtered":84823,
      "unfiltered":84823,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"d775f922-4939-451a-959b-a5523af418cf",
      "device":"LimiTTer",
      "dateString":"2018-04-04T20:57:38.475+0200",
      "sysTime":"2018-04-04T20:57:38.475+0200",
      "date":1522868258475,
      "sgv":56,
      "delta":-2.940,
      "direction":"Flat",
      "noise":1,
      "filtered":78941,
      "unfiltered":78941,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"8da9de1f-9ac1-446d-b03b-772847ee733b",
      "device":"LimiTTer",
      "dateString":"2018-04-04T20:50:38.475+0200",
      "sysTime":"2018-04-04T20:50:38.475+0200",
      "date":1522867838475,
      "sgv":66,
      "delta":-0.976,
      "direction":"Flat",
      "noise":0,
      "filtered":74941,
      "unfiltered":74941,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"2563d45c-57d9-486a-9e02-5589e88436a3",
      "device":"LimiTTer",
      "dateString":"2018-04-04T20:45:38.475+0200",
      "sysTime":"2018-04-04T20:45:38.475+0200",
      "date":1522867538475,
      "sgv":70,
      "delta":-0.976,
      "direction":"Flat",
      "noise":0,
      "filtered":80470,
      "unfiltered":80470,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"d1958b25-d23e-414a-ad3a-dc4018b58cae",
      "device":"LimiTTer",
      "dateString":"2018-04-04T20:40:38.475+0200",
      "sysTime":"2018-04-04T20:40:38.475+0200",
      "date":1522867238475,
      "sgv":75,
      "delta":-0.976,
      "direction":"Flat",
      "noise":0,
      "filtered":86235,
      "unfiltered":86235,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"f5b1672b-649a-4c95-bda4-f8a4683a1d6b",
      "device":"LimiTTer",
      "dateString":"2018-04-04T20:35:38.475+0200",
      "sysTime":"2018-04-04T20:35:38.475+0200",
      "date":1522866938475,
      "sgv":78,
      "delta":-0.976,
      "direction":"Flat",
      "noise":0,
      "filtered":89999,
      "unfiltered":89999,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"1eedef68-9bf4-48bf-9f58-858e0d506ed0",
      "device":"LimiTTer",
      "dateString":"2018-04-04T20:27:33.257+0200",
      "sysTime":"2018-04-04T20:27:33.257+0200",
      "date":1522866453257,
      "sgv":74,
      "delta":15.453,
      "direction":"SingleUp",
      "noise":1,
      "filtered":99411,
      "unfiltered":99411,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"e1568dad-72a9-40ef-985e-311814361e08",
      "device":"LimiTTer",
      "dateString":"2018-04-04T20:18:52.090+0200",
      "sysTime":"2018-04-04T20:18:52.090+0200",
      "date":1522865932090,
      "sgv":47,
      "delta":0.813,
      "direction":"Flat",
      "noise":1,
      "filtered":68352,
      "unfiltered":68352,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"e6f317f8-520b-4fd7-adcd-6a1aab2bdeae",
      "device":"LimiTTer",
      "dateString":"2018-04-04T20:13:52.090+0200",
      "sysTime":"2018-04-04T20:13:52.090+0200",
      "date":1522865632090,
      "sgv":46,
      "delta":19.422,
      "direction":"DoubleUp",
      "noise":1,
      "filtered":67411,
      "unfiltered":67411,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"6e359540-286e-4c79-88f9-038736c74e39",
      "device":"LimiTTer",
      "dateString":"2018-04-04T20:08:52.090+0200",
      "sysTime":"2018-04-04T20:08:52.090+0200",
      "date":1522865332090,
      "sgv":27,
      "delta":-5.288,
      "direction":"FortyFiveDown",
      "noise":1,
      "filtered":44941,
      "unfiltered":44941,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"3fdae952-9d81-4eec-8b7a-bc83f0e37a09",
      "device":"LimiTTer",
      "dateString":"2018-04-04T20:03:52.090+0200",
      "sysTime":"2018-04-04T20:03:52.090+0200",
      "date":1522865032090,
      "sgv":32,
      "delta":-8.600,
      "direction":"FortyFiveDown",
      "noise":1,
      "filtered":51058,
      "unfiltered":51058,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"07a705dc-f222-4527-9444-35c8c42ed215",
      "device":"LimiTTer",
      "dateString":"2018-04-04T19:54:52.090+0200",
      "sysTime":"2018-04-04T19:54:52.090+0200",
      "date":1522864492090,
      "sgv":83,
      "delta":-0.976,
      "direction":"Flat",
      "noise":0,
      "filtered":97294,
      "unfiltered":97294,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"b5e2197f-961f-4926-b2c5-a9e394a118f1",
      "device":"LimiTTer",
      "dateString":"2018-04-04T19:49:52.090+0200",
      "sysTime":"2018-04-04T19:49:52.090+0200",
      "date":1522864192090,
      "sgv":100,
      "delta":-0.976,
      "direction":"Flat",
      "noise":0,
      "filtered":119411,
      "unfiltered":119411,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"a02fbd61-cd23-4c2e-b78c-6a22715eddc5",
      "device":"LimiTTer",
      "dateString":"2018-04-04T19:44:52.090+0200",
      "sysTime":"2018-04-04T19:44:52.090+0200",
      "date":1522863892090,
      "sgv":115,
      "delta":-0.976,
      "direction":"Flat",
      "noise":0,
      "filtered":138235,
      "unfiltered":138235,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"f788cf31-821d-4495-bdfd-6a22ffe1a96f",
      "device":"LimiTTer",
      "dateString":"2018-04-04T19:39:52.090+0200",
      "sysTime":"2018-04-04T19:39:52.090+0200",
      "date":1522863592090,
      "sgv":127,
      "delta":-0.976,
      "direction":"Flat",
      "noise":0,
      "filtered":154235,
      "unfiltered":154235,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"6ab92088-f1fc-4376-b86c-4b4c95ad05ca",
      "device":"LimiTTer",
      "dateString":"2018-04-04T19:34:52.090+0200",
      "sysTime":"2018-04-04T19:34:52.090+0200",
      "date":1522863292090,
      "sgv":138,
      "delta":-0.976,
      "direction":"Flat",
      "noise":0,
      "filtered":168470,
      "unfiltered":168470,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"123ab45f-7474-40d4-9aad-128456638d65",
      "device":"LimiTTer",
      "dateString":"2018-04-04T19:29:52.090+0200",
      "sysTime":"2018-04-04T19:29:52.090+0200",
      "date":1522862992090,
      "sgv":147,
      "delta":-0.976,
      "direction":"Flat",
      "noise":0,
      "filtered":181176,
      "unfiltered":181176,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"80615492-7591-4f77-903a-cbd260d67269",
      "device":"LimiTTer",
      "dateString":"2018-04-04T19:24:52.090+0200",
      "sysTime":"2018-04-04T19:24:52.090+0200",
      "date":1522862692090,
      "sgv":156,
      "delta":-0.976,
      "direction":"Flat",
      "noise":0,
      "filtered":192235,
      "unfiltered":192235,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"cfc35281-aeed-4452-8688-4a92ac045d60",
      "device":"LimiTTer",
      "dateString":"2018-04-04T19:19:52.090+0200",
      "sysTime":"2018-04-04T19:19:52.090+0200",
      "date":1522862392090,
      "sgv":163,
      "delta":-0.976,
      "direction":"Flat",
      "noise":0,
      "filtered":201411,
      "unfiltered":201411,
      "rssi":100,
      "type":"sgv"
   },
   {  
      "_id":"4cb4166c-2ec5-4944-ac1f-fe4bdfe4448f",
      "device":"LimiTTer",
      "dateString":"2018-04-04T19:14:52.090+0200",
      "sysTime":"2018-04-04T19:14:52.090+0200",
      "date":1522862092090,
      "sgv":168,
      "delta":-0.976,
      "direction":"Flat",
      "noise":0,
      "filtered":207999,
      "unfiltered":207999,
      "rssi":100,
      "type":"sgv"
   }
]
```

**Comment by edotassi** [comment]
Hi, yes, can you explain better what data you would like to see on the watch?

**Comment by edotassi** [comment]
ok maybe i understand, do you want this block of data in a pace's springboard?

```json
{  
      "_id":"4cb4166c-2ec5-4944-ac1f-fe4bdfe4448f",
      "device":"LimiTTer",
      "dateString":"2018-04-04T19:14:52.090+0200",
      "sysTime":"2018-04-04T19:14:52.090+0200",
      "date":1522862092090,
      "sgv":168,
      "delta":-0.976,
      "direction":"Flat",
      "noise":0,
      "filtered":207999,
      "unfiltered":207999,
      "rssi":100,
      "type":"sgv"
   }
```

**Comment by Klaus3d3** [comment]
Usually its Blood glucose "sgv", the delta "delta", the "direction" shown as a trend arrow and the time since this last reading in form the "date" timestamp-actually tinesinceepoch.

Usually it looks something like this: https://m.imgur.com/5SzW3AK

I think having a curve is overdone since xdrips webservice only provides the last 15 minutes or so.

Just for info, i tried installing the micro apk of xdrip wich is for android wear and strangely i could choose xdrips integrated watchfaces on my stratos. Of course there was no low power watchface and no values shown but the watchfaces it self seem to work. Time and battery status were right. :-D maybe we could also work with this.

**Comment by Klaus3d3** [comment]
Yes, you were faster. The first data set is always the newest and the relevant.

I'm pretty shure i could do the springboard part with a lot of time, effort and try and error by myself (and google :-D), but i have no idea for the phone to watch communication. I tried so many bluetooth apps and frameworks for e. g.  a simple chat. None of them worked. With once i could send text from watch to phone but not the other way.

**Comment by edotassi** [comment]
I'll take care of bluetooth communication 😋

**Comment by Klaus3d3** [comment]
Super, let me know what i can help with.

**Comment by Klaus3d3** [comment]
Btw, in case of bad values icm used to have an Alarm. What do you think would be the best way to do so? Within the springboard/watch app or the companion. I'm thinking about something like the incoming call notification...something i have to dismiss by will.

**Comment by edotassi** [comment]
good news!
Transport service seems to work:

```
D/Amazmod ( 1004): BOOT_COMPLETED
D/Amazmod ( 1004): MainService started
D/Amazmod ( 1004): connecting companionTransporter to transportService
D/Amazmod ( 1004): action: nightscout_sync, module: com.edotassi.amazmod
D/Amazmod ( 1004): date: 1522869796436, sgv: 78, delta: -1.102, direction: Flat, noise: 1, filtered: 103764, unfiltered: 103764, rssi: 100
```

this log arrives directly from the watch 😄

**Comment by edotassi** [comment]
what are the most important data to show?

**Comment by Klaus3d3** [comment]
Usualy its:

Sgv(most important the actual glucose value)
Delta (value change since last messurement) 
Direction (string translated into an trend arrow) 
Date (in form of "xx min ago" meens date - timesinceepoch)

**Comment by Klaus3d3** [comment]
Would be good if one could enter the url in the amazmod app. 
Using the local webservice of xdrip is best cause it need no real internet connection.
But many upload their data to their nightscout site. 
For example when the diabetic kid is school, it uploads the data to nightscout and the parents can download them and have them on their watch.
The good thing is, the xdrip webservice and the nightscout site deliver the same json formated data. So you could use it local with xdrip, or remote with nightscout.

**Comment by Klaus3d3** [comment]
I've sent you my nightscout adress on xda pm for testing

**Comment by edotassi** [comment]
first tests with the UI

![senzanome](https://user-images.githubusercontent.com/3192179/38525702-9590943a-3c53-11e8-9528-497e1b45a7be.png)

**Comment by GreatApo** [comment]
I can make a widget showing the values, and maybe vibrating when refreshing or when you are out of limits. 
@edotassi can you provide me the java code getting the values?

**Comment by Klaus3d3** [comment]
Hi Apo, yes some settings would be great.

I don't know how familiar you are with this desease.

Settings wich would be usefull:
-Units mg/dl or mmol/l.  (100mg/dl ≈ 5.5 mmol/l) depends on country
-LOW and HIGH Limits for changing font color
-yes, implementing a warning would be very usefull.

**Comment by edotassi** [comment]
@GreatApo ok i will publish source code and example in new repository today

**Comment by edotassi** [comment]
@Klaus3d3 good news!

tomorrow I will send you the apk for the phone and for the clock, the data you see in the image come from your private URL 👍 

@GreatApo , I need to fix the code quickly and then I'll load the project tomorrow

![20180410_233521](https://user-images.githubusercontent.com/3192179/38585112-fe20d54e-3d18-11e8-81d4-b67727238c98.jpg)

**Comment by Klaus3d3** [comment]
That's an awesome amazmod. :-D

**Comment by edotassi** [comment]
@Klaus3d3 in the live data you provided me I do not see the delta

**Comment by Klaus3d3** [comment]
Oh, you're right. I didn't notice the difference. 
The local webservice of xdrip provides the delta. 
Can you make it being blank, when this attribute is missing?

**Comment by edotassi** [comment]
@Klaus3d3 do you have root on watch?

**Comment by Klaus3d3** [comment]
these are already the basics.

I would most of this do inside amazmod to keep the watch app as slick as possible

- A selector for mmol/mgdl (mgdl needs to be devided by 18.0182) then sending just the new value. 
- An additional menu for high/low limits for warnings
- Maybe extra limits for very low/low/good/high/very high to change the color of the value in the widget. For this i would dump one of the useless data entries like RSSI and send the color code instead.

With the warnings i'm still not sure whats the best way. Sending also the limits to the watch and let the watch decide and trigger an alarm/notification or doing this also in amazmod and sending something like the incoming call notification.

**Comment by edotassi** [comment]
@GreatApo it's simpler set this options in AmazMod, and then send to the watch

**Comment by Klaus3d3** [comment]
@edotassi maybe a stupid question, cause i think many would had it done before.
Now that we can send data to the watch...Is it possible that a watchface can also access your service and show the data?

**Comment by GreatApo** [comment]
It may be posible in both modes but watchface apks are not stable, because most of them are not open source and the libraries are not exactly the same for each version.

**Comment by Klaus3d3** [comment]
Hmm, maybe some day...just a simple one, time, date and the glucose stuff..

**Comment by jdelacasa** [comment]
hi, any update?

**Comment by Klaus3d3** [comment]
Widget is running stable with Amazmod. Nothing about watchface integration, yet. To busy.

**Comment by jdelacasa** [comment]
ok thx!, Are there the apk or do you have to compile them?

**Comment by edotassi** [comment]
@jdelacasa the service is not yet published, if you want to try it you can find it here: https://github.com/edotassi/AmazModService

ps: in the "Release" section there is an apk, it works but it is not really the last version

**Comment by Klaus3d3** [comment]
You also need the AmazMod apk with Nightscout support.

**Comment by edotassi** [comment]
for the moment I close this issue

**Comment by edotassi** [comment]
for the moment I close this issue

**Comment by sulkaharo** [comment]
Hi! Did this get published? I can't find a trace of Nightscout in Amazmod (or docs on how to install this) and would love to contribute to improve the watchface (I'm the lead dev on Nightscout).

**Comment by lfom** [comment]
> Hi! Did this get published? I can't find a trace of Nightscout in Amazmod (or docs on how to install this) and would love to contribute to improve the watchface (I'm the lead dev on Nightscout).

Hello, there. AFAIK, the support was moved to GreatFit (watchface) by @GreatApo a long time ago. AmazMod only sends the data to the watch, but I am not sure if the current version has it working. I didn't work on this ever, only Edo (that isn't working in the code anymore since a long time too) and GreatApo.

**Comment by GreatApo** [comment]
@sulkaharo, this was working in the past, but, in the end Klaus modified GreatFit and xdrip and a working watchface-widget was created (it probably still works).

Currently, Amazmod doesn't forward any data to the watch, however, in order to revive some of this features I was thinking of trying to grab the diabetic data from phone (I think xdrip has a feature that transmits the data internally in phone). Then fix the xdrip widget of GreatFit and make/enable a display menu in amazmod too.

One of the problems is that it is not easy to test these since we don't have values/updates feed.

To sum up, the current state of the feature is "not working". However we are open to ideas and suggestions that can help many people :)

**Comment by Klaus3d3** [comment]
Hi, yes my watchface/widget is still working, but the data are coming now directly from xdrip. So, if xdrip is set up as a nightscout follower, this should also work. But has nothing to do with amazmod anymore.

**Comment by GreatApo** [comment]
> Hi, yes my watchface/widget is still working, but the data are coming now directly from xdrip. So, if xdrip is set up as a nightscout follower, this should also work. But has nothing to do with amazmod anymore.

I know you had xDrip modified but you or somebody mentioned that your changes are also integrated in the official app? Can your transport method be used to grab the data from the official app?

**Comment by Klaus3d3** [comment]
Nope, if i'm right there is no nightscout app. Nightscout is a webside where you can upload your data to or download from via JSON.
Thats why you can set up xdrip to sync with any nightscout side and send it to my watchface.

If you want to integrate nightscout support into greatfit/amazmod, all you need is a field where you can enter the nightscout url and the passkey. Thats all you need. The data is within the JSON you get.

**Comment by GreatApo** [comment]
> Nope, if i'm right there is no nightscout app. Nightscout is a webside where you can upload your data to or download from via JSON.
> Thats why you can set up xdrip to sync with any nightscout side and send it to my watchface.
> 
> If you want to integrate nightscout support into greatfit/amazmod, all you need is a field where you can enter the nightscout url and the passkey. Thats all you need. The data is within the JSON you get.

Oh, ok good to know, that is easy to be done :)

**Comment by sulkaharo** [comment]
Hey related question - do any of the amazmod in-watch services work if a phone is hooked to an iOS device? I'm wondering if there's a way to produce a generic watchface that runs on an affordable watch like the Pace, which would then have the ability to load the JSON files regardless of the user's phone of choice. (I'm the lead dev on the Nightscout website side and there hasn't been a good solution for the community after Pebble disappeared)

**Comment by GreatApo** [comment]
> Hey related question - do any of the amazmod in-watch services work if a phone is hooked to an iOS device? I'm wondering if there's a way to produce a generic watchface that runs on an affordable watch like the Pace, which would then have the ability to load the JSON files regardless of the user's phone of choice. (I'm the lead dev on the Nightscout website side and there hasn't been a good solution for the community after Pebble disappeared)

Only for Android. The only way to do it for iOS is to either port the AmazfitInternetCompanion app to iOS or rely to the wifi connection of the watch, which i don't think is worth.

Now for android @Klaus3d3 did and amazing work connecting everything, check his repository, as far as I know he has a working watchface and a widget.

However since a lot of people use Amazmod, we can also re-set up this feature and I can also fix it in GreatFit.
